gensim/models/word2vec.py

```diff
@@ -1819,18 +1820,31 @@ def predict_output_word(self, context_words_list, topn=10):
         if not hasattr(self.wv, 'vectors') or not hasattr(self, 'syn1neg'):
             raise RuntimeError("Parameters required for predicting the output words not found.")
 
-        word2_indices = [self.wv.get_index(w) for w in context_words_list if w in self.wv]
-        if not word2_indices:
-            logger.warning("All the input context words are out-of-vocabulary for the current model.")
-            return None
+        if all(isinstance(w, int) for w in context_words_list):
+            # then, indices were passed. Check they are valid
+            word2_indices = np.array(context_words_list)
+            if np.any(word2_indices < 0):
+                logger.warning("All input context word indices must be non-negative.")
+                return None
+            # take only the ones in the vocabulary
+            word2_indices = word2_indices[word2_indices < self.wv.vectors.shape[0]]
+            if word2_indices.size == 0:
+                logger.warning("All the input context words are out-of-vocabulary for the current model.")
+                return None
+        else:
+            # then, words were passed. Retrieve their indices
+            word2_indices = [self.wv.get_index(w) for w in context_words_list if w in self.wv]
+            if not word2_indices:
```

Comment:
How would it affect the performance if you put the "integer indices" check after the "string indices"?
That is, "if strs, convert to ints => continue always with test ints, test ints for empty, etc". It would unify the two code paths, no need to test twice.

Reply:
Or indeed, like @gojomo says, unify even more and accept mixed strings-and--ints, in a single code path.

Reply:
I don't know about performance but less duplicates sounds righter.

Reply:
I do think the most-compact & flexible code would accept a list-of-words-or-word-int-indexes, then convert it to all indexes (by replacing any strings with ints, in a single list pass). 

If any bad indexes are passed, or words *not* in the model, whatever errors those unchecked conditions trigger are probably clear enough: `IndexError` or `KeyError`, with a stack through the caller code, and this method's code, that uses a bad parameter. So essentially no overhead of parameter-checking. 

I'm especially willing to consider this low-overhead approach because I'd categorize this method as 'advanced' and 'experimental' - it's *not* a typical use/function of `Word2Vec`. The method only works for one mode (negative-sampling), and doesn't (without weighting) truly approximate what a training-time negative-sampling prediction would be. The need to sort the activation values of *all* potential vocabulary words before returning the top-N means it's an inherently fairly-expensive operation. As such, it already has many caveats (that should be reflected in its doc-comment) - and the extra expectation that it should only be called with sensible arguments is no extra burden. 

But, also, that `KeyedVectors.get_index()` will already take a string-or-int, and return an int index:

https://github.com/RaRe-Technologies/gensim/blob/29fecbfc38a0f0f0783fe80e674f008a2620da30/gensim/models/keyedvectors.py#L383

So I think in fact choosing to use that `.get_index()` method would ensure a single-pass over the word-list would always through a descriptive `KeyError` if *either* the string-word isn't present, or the int-index is out-of-range.